File: frontend/js/verifier.js
class FrontendVerifier {
    constructor() {
        this.verificationQueue = [];
        this.isVerifying = false;
        this.trustedKeys = [];
        this.loadTrustedKeys();
    }

    async loadTrustedKeys() {
        try {
            const response = await fetch('/api/trusted_keys');
            const data = await response.json();
            this.trustedKeys = data.keys;
            console.log(`Loaded ${this.trustedKeys.length} trusted keys`);
        } catch (error) {
            console.error('Error loading trusted keys:', error);
        }
    }

    async verifyData(ipfsHash) {
        try {
            // Add to verification queue
            this.verificationQueue.push(ipfsHash);
            
            // Start verification process if not already running
            if (!this.isVerifying) {
                this.isVerifying = true;
                await this.processVerificationQueue();
            }
            
            return true;
        } catch (error) {
            console.error('Error adding to verification queue:', error);
            return false;
        }
    }

    async processVerificationQueue() {
        while (this.verificationQueue.length > 0) {
            const ipfsHash = this.verificationQueue.shift();
            
            try {
                // First, try client-side verification if we have trusted keys
                if (this.trustedKeys.length > 0) {
                    const data = await this.fetchIpfsData(ipfsHash);
                    if (data && data.signature) {
                        // In a real implementation, we would use WebCrypto API here
                        // For now, we'll use the placeholder implementation
                        const isVerified = await this.verifySignatureClientSide(data);
                        if (isVerified) {
                            this.updateVerificationStatus(ipfsHash, 'verified');
                            continue; // Skip backend verification if client-side succeeded
                        }
                    }
                }
                
                // Fall back to backend verification
                const response = await fetch('/api/verify', {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    body: JSON.stringify({ ipfs_hash: ipfsHash })
                });
                
                const result = await response.json();
                
                if (result.verified) {
                    // Update UI to show verified status
                    this.updateVerificationStatus(ipfsHash, 'verified');
                } else {
                    // Update UI to show unverified status
                    this.updateVerificationStatus(ipfsHash, 'unverified');
                }
            } catch (error) {
                console.error('Verification failed for', ipfsHash, error);
                this.updateVerificationStatus(ipfsHash, 'error');
            }
        }
        
        this.isVerifying = false;
    }

    async fetchIpfsData(ipfsHash) {
        try {
            // Fetch data from IPFS via backend API
            const response = await fetch(`/api/verify`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({ ipfs_hash: ipfsHash })
            });
            
            const result = await response.json();
            if (result.verified === undefined) {
                // This is the raw data response
                return { ipfs_hash: ipfsHash, signature: result.signature, data: result.data };
            }
            return null;
        } catch (error) {
            console.error('Error fetching IPFS data:', error);
            return null;
        }
    }

    async verifySignatureClientSide(data) {
        try {
            // Extract signature and data without signature
            const signature = data.signature;
            const dataCopy = { ...data };
            delete dataCopy.signature;
            
            // Check if we have trusted keys
            if (!this.trustedKeys || this.trustedKeys.length === 0) {
                console.log('No trusted keys available for verification');
                return false;
            }
            
            // For each trusted key, try to verify the signature
            for (const publicKey of this.trustedKeys) {
                const isValid = await verifySignatureWithWebCrypto(dataCopy, signature, publicKey);
                if (isValid) {
                    return true;
                }
            }
            
            return false;
        } catch (error) {
            console.error('Client-side signature verification failed:', error);
            return false;
        }
    }
    updateVerificationStatus(ipfsHash, status) {
        // Update the UI elements for this IPFS hash
        const elements = document.querySelectorAll(`[data-ipfs-hash="${ipfsHash}"]`);
        elements.forEach(element => {
            element.classList.remove('status-pending', 'status-verified', 'status-unverified', 'status-error');
            element.classList.add(`status-${status}`);
            
            // Update status text
            const statusElement = element.querySelector('.verification-status');
            if (statusElement) {
                statusElement.textContent = this.getStatusText(status);
            }
        });
    }

    getStatusText(status) {
        switch (status) {
            case 'verified': return 'Verified';
            case 'unverified': return 'Unverified';
            case 'error': return 'Verification Error';
            default: return 'Pending Verification';
        }
    }

    async verifyAllDisplayedContent() {
        // Find all IPFS hashes currently displayed
        const contentElements = document.querySelectorAll('[data-ipfs-hash]');
        const hashes = Array.from(contentElements).map(el => el.getAttribute('data-ipfs-hash'));
        
        // Verify each hash
        for (const hash of hashes) {
            await this.verifyData(hash);
        }
    }
}

// Helper function to sort object keys
function sortKeys(key, value) {
    if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
        return Object.keys(value).sort().reduce((result, sortedKey) => {
            result[sortedKey] = value[sortedKey];
            return result;
        }, {});
    }
    return value;
}

// WebCrypto verification implementation
async function verifySignatureWithWebCrypto(data, signatureB64, publicKeyPem) {
    try {
        // Remove signature from data for verification
        const dataCopy = { ...data };
        delete dataCopy.signature;
        
        // Convert data to JSON string for verification
        const json_data = new TextEncoder().encode(JSON.stringify(dataCopy, sortKeys));
        
        // Decode signature
        const signature = base64ToUint8Array(signatureB64);
        
        // Load public key
        const publicKey = await crypto.subtle.importKey(
            'spki',
            pemToUint8Array(publicKeyPem),
            {
                name: 'RSASSA-PKCS1-v1_5',
                hash: { name: 'SHA-256' }
            },
            false,
            ['verify']
        );
        
        // Verify signature
        const isValid = await crypto.subtle.verify(
            {
                name: 'RSASSA-PKCS1-v1_5'
            },
            publicKey,
            signature,
            json_data
        );
        
        return isValid;
    } catch (error) {
        console.error('WebCrypto signature verification failed:', error);
        return false;
    }
}

// Helper function to convert Base64 string to Uint8Array
function base64ToUint8Array(base64) {
    const binaryString = atob(base64);
    const len = binaryString.length;
    const bytes = new Uint8Array(len);
    for (let i = 0; i < len; i++) {
        bytes[i] = binaryString.charCodeAt(i);
    }
    return bytes;
}

// Helper function to convert PEM to Uint8Array
function pemToUint8Array(pem) {
    // Remove header, footer, and newlines
    const pemHeader = "-----BEGIN PUBLIC KEY-----";
    const pemFooter = "-----END PUBLIC KEY-----";
    const pemContents = pem.substring(pemHeader.length, pem.length - pemFooter.length).replace(/\n/g, '');
    
    // Convert from Base64 to Uint8Array
    return base64ToUint8Array(pemContents);
}

// Initialize verifier when page loads
document.addEventListener('DOMContentLoaded', () => {
    window.frontendVerifier = new FrontendVerifier();
    
    // Verify all content on page load
    window.frontendVerifier.verifyAllDisplayedContent();
});

File: verifier.py
import json
import base64
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives import serialization
import ipfshttpclient

class DataVerifier:
    def __init__(self, trusted_keys_file='trusted_keys.json'):
        """Initialize verifier with trusted public keys"""
        self.trusted_keys_file = trusted_keys_file
        self.trusted_keys = self.load_trusted_keys()
        self.ipfs_client = ipfshttpclient.connect('/ip4/127.0.0.1/tcp/5001')
    
    def load_trusted_keys(self):
        """Load trusted public keys from JSON file"""
        try:
            with open(self.trusted_keys_file, 'r') as f:
                data = json.load(f)
                return [key.encode('utf-8') for key in data['keys']]
        except FileNotFoundError:
            print(f"Warning: {self.trusted_keys_file} not found. No trusted keys loaded.")
            return []
        except Exception as e:
            print(f"Error loading trusted keys: {e}")
            return []
    
    def verify_signature(self, data, signature_b64, public_key_pem):
        """Verify signature using a specific public key"""
        try:
            # Remove signature from data for verification
            data_copy = data.copy()
            if 'signature' in data_copy:
                del data_copy['signature']
            
            # Convert data to JSON string for verification
            json_data = json.dumps(data_copy, sort_keys=True).encode('utf-8')
            
            # Decode signature
            signature = base64.b64decode(signature_b64)
            
            # Load public key
            public_key = serialization.load_pem_public_key(public_key_pem)
            
            # Verify signature
            public_key.verify(
                signature,
                json_data,
                padding.PKCS1v15(),
                hashes.SHA256()
            )
            return True
        except Exception as e:
            # Signature verification failed
            return False
    
    def verify_ipfs_data(self, ipfs_hash):
        """Verify data from IPFS using trusted keys"""
        try:
            # Get data from IPFS
            data = self.ipfs_client.get_json(ipfs_hash)
            
            # Check if data has signature
            if 'signature' not in data:
                print(f"Data {ipfs_hash} has no signature")
                return False
            
            # Try to verify with each trusted key
            signature = data['signature']
            for public_key_pem in self.trusted_keys:
                if self.verify_signature(data, signature, public_key_pem):
                    print(f"Data {ipfs_hash} verified with trusted key")
                    return True
            
            # No trusted key could verify the data
            print(f"Data {ipfs_hash} signature not verified by any trusted key")
            return False
            
        except Exception as e:
            print(f"Error verifying IPFS data {ipfs_hash}: {e}")
            return False
    
    def add_trusted_key(self, public_key_pem):
        """Add a new trusted public key"""
        if isinstance(public_key_pem, str):
            public_key_pem = public_key_pem.encode('utf-8')
            
        if public_key_pem not in self.trusted_keys:
            self.trusted_keys.append(public_key_pem)
            self.save_trusted_keys()
            print("Trusted key added successfully")
        else:
            print("Key already in trusted list")
    
    def save_trusted_keys(self):
        """Save trusted keys to JSON file"""
        try:
            keys_data = {
                'keys': [key.decode('utf-8') if isinstance(key, bytes) else key for key in self.trusted_keys]
            }
            with open(self.trusted_keys_file, 'w') as f:
                json.dump(keys_data, f, indent=2)
        except Exception as e:
            print(f"Error saving trusted keys: {e}")

# Example usage
if __name__ == "__main__":
    verifier = DataVerifier()
    
    # Example: verify data from IPFS
    # result = verifier.verify_ipfs_data('QmExampleHash')
    # print(f"Verification result: {result}")
    
    # Example: add a new trusted key
    # with open('new_key.pub', 'r') as f:
    #     new_key = f.read()
    # verifier.add_trusted_key(new_key)